### PR TITLE
drivers/at86rf215: implement MR-OFDM

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This adds MR-OFDM modulation to the `at86rf215` driver.

This also adds new commands to `ifconfig` to configure the new modulation parameters.

### Testing procedure

Switch to the new PHY mode with `ifconfig 7 set phy_mode mr-ofdm`. Note that the `L2-PDU` has changed to 2022 bytes (802.15.4g can send 2047 byte frames).

```
2020-05-04 03:41:16,173 # Iface  7  HWaddr: 01:57  Channel: 0  NID: 0x23  PHY: MR-OFDM 
2020-05-04 03:41:16,180 #            Option: 1  MCS: 0 (BPSK, rate 1/2, 4x frequency repetition) 
2020-05-04 03:41:16,184 #           Long HWaddr: 1A:F9:8F:5F:30:5D:05:8E 
2020-05-04 03:41:16,191 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-05-04 03:41:16,197 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:2022 MTU:1280  HL:64  RTR  
2020-05-04 03:41:16,199 #           6LO  IPHC  
2020-05-04 03:41:16,202 #           Source address length: 8
2020-05-04 03:41:16,204 #           Link type: wireless
2020-05-04 03:41:16,210 #           inet6 addr: fe80::18f9:8f5f:305d:58e  scope: link  VAL
2020-05-04 03:41:16,213 #           inet6 group: ff02::2
2020-05-04 03:41:16,216 #           inet6 group: ff02::1
2020-05-04 03:41:16,219 #           inet6 group: ff02::1:ff5d:58e
2020-05-04 03:41:16,220 #           
2020-05-04 03:41:16,223 #           Statistics for Layer 2
2020-05-04 03:41:16,226 #             RX packets 0  bytes 0
2020-05-04 03:41:16,230 #             TX packets 4 (Multicast: 4)  bytes 172
2020-05-04 03:41:16,233 #             TX succeeded 4 errors 0
2020-05-04 03:41:16,236 #           Statistics for IPv6
2020-05-04 03:41:16,239 #             RX packets 0  bytes 0
2020-05-04 03:41:16,244 #             TX packets 4 (Multicast: 4)  bytes 256
2020-05-04 03:41:16,247 #             TX succeeded 4 errors 0
2020-05-04 03:41:16,247 # 
2020-05-04 03:41:30,034 #  ping6 fe80::2068:3123:59f5:d23a%7
2020-05-04 03:41:30,053 # 12 bytes from fe80::2068:3123:59f5:d23a%7: icmp_seq=0 ttl=64 rssi=-41 dBm time=12.051 ms
2020-05-04 03:41:31,054 # 12 bytes from fe80::2068:3123:59f5:d23a%7: icmp_seq=1 ttl=64 rssi=-40 dBm time=12.051 ms
2020-05-04 03:41:32,054 # 12 bytes from fe80::2068:3123:59f5:d23a%7: icmp_seq=2 ttl=64 rssi=-41 dBm time=12.051 ms
```

You should be able to configure the OFDM option (`ifconfig 7 set option [1…4]`) as well as the Modulation and Coding Scheme (`ifconfig 7 set mcs [0…6]`):

```
2020-05-04 03:49:00,938 #  ifconfig 7 set mcs 2
2020-05-04 03:49:04,010 #  ifconfig 7 set option 4
2020-05-04 03:52:49,320 #  ping6 fe80::2068:3123:59f5:d23a%7
2020-05-04 03:52:49,348 # 12 bytes from fe80::2068:3123:59f5:d23a%7: icmp_seq=0 ttl=64 rssi=-42 dBm time=19.828 ms
2020-05-04 03:52:50,348 # 12 bytes from fe80::2068:3123:59f5:d23a%7: icmp_seq=1 ttl=64 rssi=-42 dBm time=19.828 ms
2020-05-04 03:52:51,348 # 12 bytes from fe80::2068:3123:59f5:d23a%7: icmp_seq=2 ttl=64 rssi=-42 dBm time=19.827 ms
2020-05-04 03:52:51,348 # 
2020-05-04 03:52:51,353 # --- fe80::2068:3123:59f5:d23a PING statistics ---
2020-05-04 03:52:51,358 # 3 packets transmitted, 3 packets received, 0% packet loss
2020-05-04 03:52:51,362 # round-trip min/avg/max = 19.827/19.827/19.828 ms
```

Nodes can talk to each other if they use the same option, but different Modulation & Coding Schemes.

### Issues/PRs references

split off #12128
requires #13979 to send full 2k frames

### CODE REVIEW

**Comment by benpicco** [comment]
@tinstructor I heard you like MR-OFDM, want to give this a try? :wink:

**Comment by tinstructor** [comment]
@benpicco I'll run through your testing procedure quickly tomorrow morning. Mind you, due to Corona lockdown measures I currently don't have access to a signal analyzer should things need troubleshooting.

**Comment by tinstructor** [comment]
@benpicco Remind me again, which example application do you typically use to test drivers like this?

**Comment by benpicco** [comment]
`examples/gnrc_networking` should work

**Comment by tinstructor** [comment]
@benpicco I've quickly done some testing with two OpenMote-B nodes and it seems to work just fine. Of course you'd probably want to run a few hundred pings for each interface / each combination of option + MCS to be sure

File: drivers/include/at86rf215.h [hunk 1]
```diff
@@ -137,6 +137,24 @@ enum {
...
+#endif
+/** @} */
+
+/**
+ * @name    Default MR-OFDM Modulation & Coding Scheme
+ * @{
+ */
+#ifndef AT86RF215_DEFAULT_MR_OFDM_MCS
+#define AT86RF215_DEFAULT_MR_OFDM_MCS    (2)
+#endif
```
Comment (fjmolinas) [inline drivers/include/at86rf215.h:155]: If these are configurations maybe we should start adding the `CONFIG_`

File: drivers/at86rf215/include/at86rf215_internal.h [hunk 2]
```diff
@@ -455,15 +513,6 @@ void at86rf215_disable_rpc(at86rf215_t *dev);
...
 void at86rf215_enable_rpc(at86rf215_t *dev);
 
-/**
- * @brief Notify the driver and stack about a change in transmission mode
- *        which may result in a change of PDU.
- *
- * @param[in] dev           device that changed it's mode
- * @param[in] new_mode      the new transmission mode
- */
-bool at86rf215_switch_mode(at86rf215_t *dev, uint8_t new_mode);
```
Comment (fjmolinas) [inline drivers/at86rf215/include/at86rf215_internal.h:465]: Could you remove this in its own commit?

File: drivers/at86rf215/at86rf215_ofdm.c [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        return -EINVAL;
+    }
+
+    at86rf215_await_state_end(dev, RF_STATE_TX);
+
+    /* disable radio */
+    at86rf215_reg_write(dev, dev->BBC->RG_PC, 0);
+
+    /* set receiver gain target according to data sheet */
+    at86rf215_reg_write(dev, dev->RF->RG_AGCS, 3 << AGCS_TGT_SHIFT);
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_ofdm.c:254]: Is that the same?
```suggestion
    at86rf215_reg_write(dev, dev->RF->RG_AGCS, AGCS_TGT_MASK);
```

Comment (benpicco) [reply drivers/at86rf215/at86rf215_ofdm.c:254]: `AGCS_TGT_MASK` is 0xE0, `3 << AGCS_TGT_SHIFT` will work out to 0x60

File: drivers/at86rf215/include/at86rf215_internal.h (hunk 2 above)
Comment (benpicco) [reply drivers/at86rf215/include/at86rf215_internal.h:465]: Move it to a separate commit.

File: drivers/at86rf215/at86rf215_ofdm.c [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+ * @brief       Configuration of the MR-OFDM PHY on the AT86RF215 chip
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include "at86rf215.h"
+#include "at86rf215_internal.h"
+
+#define ENABLE_DEBUG (1)
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_ofdm.c:23]:
```suggestion
#define ENABLE_DEBUG (0)
```

**Comment by fjmolinas** [comment]
> @benpicco I've quickly done some testing with two OpenMote-B nodes and it seems to work just fine. Of course you'd probably want to run a few hundred pings for each interface / each combination of option + MCS to be sure

Your application that looped over the different configurations would be very usefull here :)

**Comment by benpicco** [comment]
> Your application that looped over the different configurations would be very usefull here :)

[That one](https://github.com/benpicco/range_test)? 
Probably needs an update to match the upstream API now.

**Comment by fjmolinas** [comment]
> > Your application that looped over the different configurations would be very usefull here :)
> 
> [That one](https://github.com/benpicco/range_test)? 
> Probably needs an update to match the upstream API now.

Did some testing mixing up configurations and it all seems to work. trying to see if I can cover them all automaticaly.

**Comment by fjmolinas** [comment]
> > > Your application that looped over the different configurations would be very usefull here :)
> > 
> > [That one](https://github.com/benpicco/range_test)? 
> > Probably needs an update to match the upstream API now.
> 
> Did some testing mixing up configurations and it all seems to work. trying to see if I can cover them all automatically.

I used https://github.com/RIOT-OS/RIOT/pull/13612 to script testing over all possible combinations. Worked for all pining both ways with `-c 100 -s 100 -i 100`. For the options/mcs allowing higher bandwith I was able to ping at much higher rates.

**Comment by fjmolinas** [comment]
@benpicco would you mind uncrustifying the files?

**Comment by benpicco** [comment]
I removed some of the crust, although IMHO that added to the taste.
Or do you mean also uncrustify the existing code?

**Comment by fjmolinas** [comment]
> I removed some of the crust, although IMHO that added to the taste.
> Or do you mean also uncrustify the existing code?

The new code, I noticed that `uncrustify -c uncrustify-riot.cfg  drivers/at86rf215/at86rf215_ofdm.c ` suggested many changes :)

**Comment by fjmolinas** [comment]
Please squash @benpicco (BTW I don't get nitified when you push a fix, so please ping me when you do so I notice :) )

File: drivers/at86rf215/at86rf215_ofdm.c [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+                        _TXDFE_SR(option) | _RXDFE_RCUT(option, superGHz));
+    /* Set Power Amplifier Ramp Time; fLPCUT */
+    at86rf215_reg_write(dev, dev->RF->RG_TXCUTC,
+                        RF_PARAMP8U | _TXCUTC_LPFCUT(option));
+    /* Set fS; fCUT for TX */
+    at86rf215_reg_write(dev, dev->RF->RG_TXDFE,
+                        _TXDFE_SR(option) | _TXDFE_RCUT(option));
+
+    /* set channel spacing */
+    at86rf215_reg_write(dev, dev->RF->RG_CS, _channel_spacing_kHz(option) / 25);
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_ofdm.c:184]: Just noticed this 25, why is it?

Comment (fjmolinas) [reply drivers/at86rf215/at86rf215_ofdm.c:184]: Oh yeah 25Khz resolution

Comment (fjmolinas) [reply drivers/at86rf215/at86rf215_ofdm.c:184]: Can you add a comment like for `o-qpsk`?

`/* set channel spacing with 25 kHz resolution */`

Comment (benpicco) [reply drivers/at86rf215/at86rf215_ofdm.c:184]: Done! And I also undid some of the damage done by `uncrustify` in the next line.

**Pull request merged**